proof_of_concept/registry.py

```diff
@@ -5,194 +5,279 @@
+        Args:
+            site: The site at which this server is located.
+            namespace: The namespace governed by this server.
+            server: The policy server.
 
-        Returns:
-            A list of names as strings.
         """
-        return list(self._runners.keys())
+        self.site = site
+        self.namespace = namespace
+        self.server = server
 
-    def get_runner(self, name: str) -> ILocalWorkflowRunner:
-        """Look up a LocalWorkflowRunner.
 
-        Args:
-            name: The name of the runner to look up.
+class Registry:
+    """Global registry of remote-accessible things.
 
-        Return:
-            The runner with that name.
+    Registers runners, stores, and assets. In a real system, runners
+    and stores would be identified by a URL, and use the DNS to
+    resolve. For now the registry helps with this.
+    """
+    def __init__(self) -> None:
+        """Create a new registry."""
+        self._assets = dict()           # type: Dict[str, str]
 
-        Raises:
-            KeyError: If no runner with the given name is
-                    registered.
-        """
-        return self._runners[name]
+        self._archive = ReplicableArchive[RegisteredObject]()
+        self._store = CanonicalStore[RegisteredObject](self._archive)
+        self.replication_server = ReplicationServer[RegisteredObject](
+                self._archive, 1.0)
 
-    def get_runner_admin(self, name: str) -> str:
-        """Look up who administrates a given runner.
+    def register_party(
+            self, name: str, namespace: str, public_key: RSAPublicKey) -> None:
+        """Register a party with the DDM.
 
         Args:
-            name: The name of the runner to look up.
+            name: Name of the party.
+            namespace: ID namespace owned by this party.
+            public_key: Public key of this party.
+        """
+        for i in self._store.objects():
+            if isinstance(i, PartyDescription) and i.name == name:
+                raise RuntimeError('There is already a party called {}'.format(
+                    name))
 
-        Return:
-            The name of the party administrating it.
+        party_desc = PartyDescription(name, public_key)
+        self._store.insert(party_desc)
+        self._store.insert(NamespaceDescription(namespace, party_desc))
+
+    def register_site(self, name: str, admin_name: str) -> None:
+        """Register a Site with the Registry.
+
+        Args:
+            name: Name of the site.
+            admin_name: Party administrating this site.
 
-        Raises:
-            KeyError: If no runner with the given name is regstered.
         """
-        return self._runner_admins[name]
+        for o in self._store.objects():
+            if isinstance(o, SiteDescription) and o.name == name:
+                raise RuntimeError('There is already a site called {}'.format(
+                    name))
```

Comment:
This pattern is repeated a lot, maybe you can replace it with a method:

```python
def _object_exists(name: str, type: Any):
    for o in self._store.objects():
        if isinstance(o, type) and o.name == name:
             return True
    return False
```

Reply:
Maybe you can define this in the CanonicalStore class.

Reply:
I considered it, but we don't know in general how the user wants to determine whether two objects are identical. In this case by the `name` attribute, but that's just how this class happens to work. You could have the user pass a lamba function, but that quickly becomes harder to understand than having a bit of repetition, IMHO. I'll have another look at it though, see if I can improve it.

Reply:
Okay, I managed, although it does use a bit of introspection. But that's Pythonic and it looks better. What do you think?

Reply:
This looks very nice actually (I guess I like Python and better looks)! Great work 👍